auth_dynamic_groups/models/res_groups.py

```diff
@@ -1,44 +1,181 @@
 # Copyright 2013-2019 Therp BV <https://therp.nl>
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+# pylint: disable=missing-docstring,invalid-name,no-self-use,protected-access
+import logging
+
 from odoo import _, api, exceptions, fields, models
 from odoo.tools.safe_eval import safe_eval
 
 
+_logger = logging.getLogger(__name__)
+
+
 class ResGroups(models.Model):
     _inherit = 'res.groups'
 
-    is_dynamic = fields.Boolean('Dynamic')
+    is_dynamic = fields.Boolean(string='Dynamic', index=True)
+    is_automatic = fields.Boolean(string='Filled programmatically')
+    enable_debug = fields.Boolean(string='Display detailed debug messages')
+
+    def _get_group_use_selection(self):
+        """Get values that determine how conditions should be used."""
+        return[
+            ('ignore', 'Do not use this condition'),
+            ('exclude', 'Exclude user from group'),
+            ('include', 'Include user in group'),
+            ('use', 'Include if all used conditions satisfied'),
+        ]
+
+    # Fields to support selection by formula.
+    dynamic_method_formula = fields.Selection(
+        selection="_get_group_use_selection",
+        string="Use formula",
+        help="Specify wether to include users based on formula")
     dynamic_group_condition = fields.Text(
-        'Condition', help='The condition to be met for a user to be a '
-        'member of this group. It is evaluated as python code at login '
-        'time, you get `user` passed as a browse record')
+        string='Condition',
+        default='False',
+        help="The condition to be met for a user to be a member"
+             " of this group.\n"
+             "It is evaluated as python code at login time,"
+             " you get `user` passed as a browse record")
+
+    # Fields to support selection by category.
+    dynamic_method_category = fields.Selection(
+        selection="_get_group_use_selection",
+        string="Use partner category",
+        help="Specify wether to include users based on partner category")
+    partner_category_ids = fields.Many2many(
+        comodel_name='res.partner.category',
+        relation='dynamic_group_partner_category',
+        column1='gid', column2='category_id',
+        string="Partner categories",
+        help="Used when selection is based on partner category.")
+
+    @api.multi
+    def action_refresh_users(self):
+        """Refresh the users in this group.
+
+        Limit this to 512 users, to prevent huge performance problems.
```

Comment:
I imagine this could cause regressions on installations with 513 users or more. Maybe this should be a global setting?